src/Task.cc

```diff
@@ -3694,6 +3694,69 @@ static void copy_mem_mapping(Task* from, Task* to, const KernelMapping& km) {
   }
 }
 
+// https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/fs/proc/task_mmu.c?h=v6.3#n1352
+#define PM_PRESENT (1ULL << 63)
+#define PM_SWAP    (1ULL << 62)
+
+static bool copy_mem_mapping_just_used(Task* from, Task* to, const KernelMapping& km)
+{
+  ScopedFd& fd = from->pagemap_fd();
+  if (!fd.is_open()) {
+    LOG(debug) << "Failed to open " << from->proc_pagemap_path();
+    return false;
+  }
+
+  size_t pagesize = page_size();
+  uint64_t pages_present = 0; // Just for logging
+
+  const int max_buf_size = 65536;
+  vector<uint64_t> buf;
+
+  for (uintptr_t offset = 0; offset < km.size() / pagesize; offset += max_buf_size) {
+    auto read_offset = (km.start().as_int() / pagesize + offset) * sizeof(uint64_t);
+    ssize_t read_count = min<ssize_t>(max_buf_size, km.size() / pagesize - offset) * sizeof(uint64_t);
+    buf.resize(read_count / sizeof(uint64_t));
+    ssize_t bytes_read = pread(fd, buf.data(), read_count, read_offset);
+    ASSERT(from, read_count == bytes_read);
+
+    // A chunk was read from pagemap above, now iterate through it to detect
+    // if memory is physically present (bit 63, PM_PRESENT) or in swap (bit 62, PM_SWAP) in Task "from".
+    // If yes, just transfer those pages to the new Task "to".
+    // Also try to find consecutive pages to copy them in one operation.
+    // The file /proc/PID/pagemap consists of 64-bit values, each describing
+    // the state of one page. See https://www.kernel.org/doc/Documentation/vm/pagemap.txt
+
+    for (size_t page = 0; page < bytes_read / sizeof(uint64_t); ++page) {
+      if (buf[page] & (PM_PRESENT | PM_SWAP)) {
+        auto start = km.start() + (offset + page) * pagesize;
+        if (start >= km.end()) {
+          break;
+        }
+        ++pages_present;
+
+        // Check for consecutive used pages
+        while (page < bytes_read / sizeof(uint64_t) - 1 &&
+               buf[page + 1] & (PM_PRESENT | PM_SWAP))
+        {
+          if (km.start() + (offset + page + 1) * pagesize >= km.end()) {
```

Comment:
Why do we need this check? Seems to me we couldn't have read more bytes than we needed for `km`?